Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h

```diff
@@ -0,0 +1,438 @@
+        }
+        else if constexpr (std::is_pointer_v<T> && std::is_class_v<std::remove_pointer_t<T>>)
+            return GetClassTypeName((T)0);
+        else if constexpr (std::is_same_v<T, dummy_type>)
+            return "";
+    }
+
+    static SString ResolveParameter(lua_State* L, std::size_t index)
+    {
+        switch (lua_type(L, index))
+        {
+            case LUA_TNUMBER:
+                return SString("number (%d)", lua_tonumber(L, index));
+            case LUA_TSTRING:
+            {
+                std::size_t iLen;
+                const char* szValue = lua_tolstring(L, index, &iLen);
+                std::string strValue(szValue, iLen);
+                if (strValue.length() > 10)
+                {
+                    // Limit to 10 characters
+                    strValue.resize(10);
+                    strValue[9] = '.';
+                    strValue[8] = '.';
+                    strValue[7] = '.';
+                }
+                // Avoid printing binary data
+                if (std::find_if(strValue.begin(), strValue.end(), [](char ch) { return !(std::isprint(ch)); }) != strValue.end())
+                    return "string";
+                else
+                {
+                    return SString("string (\"%s\")", strValue.c_str());
+                }
+            }
+            case LUA_TBOOLEAN:
+                return SString("boolean (%s)", lua_toboolean(L, index) == 1 ? "true" : "false");
+            case LUA_TNIL:
+                return "nil";
+            case LUA_TNONE:
+                return "none";
+            case LUA_TTABLE:
+                return "table";
+            case LUA_TFUNCTION:
+                return "function";
+            case LUA_TTHREAD:
+                return "coroutine";
+            case LUA_TUSERDATA:
+                return GetUserDataClassName(*((void**)lua_touserdata(L, index)), L);
+            case LUA_TLIGHTUSERDATA:
+                return GetUserDataClassName(lua_touserdata(L, index), L);
+        }
+        return "";
+    }
+
+    // Pop should remove a T from the Lua Stack after verifying that it is a valid type
+    // Pop may also throw a LuaArgumentError to indicate failure
+    template <typename T>
+    inline T Pop(lua_State* L, std::size_t& index)
+    {
+        if (!TypeMatch<T>(L, index))
+        {
+            SString strReceived = ResolveParameter(L, index);
+            SString strExpected = typeToName<T>();
+            SString strMessage("Expected %s at argument %d, got %s", strExpected.c_str(), index, strReceived.c_str());
+            strError = strMessage;
+            return T{};
+        }
+        return PopUnsafe<T>(L, index);
+    }
+
+    // Special type matcher for variants. Returns -1 if the type does not match
+    // returns n if the nth type of the variant matches
+    template <typename T>
+    inline int TypeMatchVariant(lua_State* L, std::size_t index)
+    {
+        if constexpr (std::is_same_v<T, std::variant<>>)
```

Comment:
```suggestion
        // Error out if the variant contains 0 types.
        if constexpr (std::is_same_v<T, std::variant<>>)
```

Consider short-circuiting in this method as well.

Nice use of recursion. Make it obvious by:
- short-circuiting
- using comments like `// Base case: when variant is empty`, `// Base case: when matching with first type in the variant`, `// General case: attempt next type in variant`